### PR TITLE
Panic on Enclave Fatal Error

### PR DESCRIPTION
**Pull request by dolanbernard**
* Removed ledger enclave check
* Added check for result of enclave calls to panic on all fatal errors

### Motivation

Any time an enclave encounters a non-recoverable error, panic the untrusted code to force a restart.

### CODE REVIEW

File: enclave-boundary/src/untrusted.rs [hunk 1]
```diff
@@ -35,27 +35,45 @@ pub fn make_variable_length_ecall(
...
+                        outbuf.truncate(outbuf_used);
+                        break Ok(outbuf);
+                    },
+                    // Fatal runtime errors
+                    sgx_status_t::SGX_ERROR_INVALID_FUNCTION |
+                        sgx_status_t::SGX_ERROR_OUT_OF_TCS |
+                        sgx_status_t::SGX_ERROR_ENCLAVE_CRASHED |
+                        sgx_status_t::SGX_ERROR_ECALL_NOT_ALLOWED |
+                        sgx_status_t::SGX_ERROR_OCALL_NOT_ALLOWED |
+                        sgx_status_t::SGX_ERROR_STACK_OVERRUN => {
```
Comment (nick-mobilecoin) [inline enclave-boundary/src/untrusted.rs:62]: I don't think it's a blocker but with the number of enums compared and copied between the two return values, suggest adding a local function like
```rust
fn is_fatal_sgx_status(status: sgx_status_t) -> bool {
    match status {
        a | b | c => true,
        _ => false,
    }
}
```

File: enclave-boundary/src/untrusted.rs [hunk 2]
```diff
@@ -35,27 +35,45 @@ pub fn make_variable_length_ecall(
...
-                &mut outbuf_retry_id,
-            )
-        } {
-            sgx_status_t::SGX_SUCCESS => match retval {
-                sgx_status_t::SGX_ERROR_OUT_OF_MEMORY => continue,
-                sgx_status_t::SGX_SUCCESS => {
-                    outbuf.truncate(outbuf_used);
-                    break Ok(outbuf);
+                &mut outbuf_retry_id
+            );
```
Comment (nick-mobilecoin) [inline enclave-boundary/src/untrusted.rs:48]: :wrench: I think you should be able to end the unsafe block here
```suggestion
            );
        }
```

**Review by nick-mobilecoin — APPROVED** [review]
LGTM some suggestions that may be nice to get in, but not blockers

File: enclave-boundary/src/untrusted.rs (hunk 2 above)
Comment (nick-mobilecoin) [reply enclave-boundary/src/untrusted.rs:48]: It may need to be
```rust
let result = unsafe { ecall_fcn(...) };
``` 
or similar

**Review by eranrund — APPROVED** [review]

File: enclave-boundary/src/untrusted.rs (hunk 2 above)
Comment (dolanbernard) [reply enclave-boundary/src/untrusted.rs:48]: Ah, yes, I like this better

File: enclave-boundary/src/untrusted.rs (hunk 1 above)
Comment (dolanbernard) [reply enclave-boundary/src/untrusted.rs]: OK, I made a commit that does this and moves that check outside of the match blocks. It should preserve the order of the original checks.

Comment (eranrund) [reply enclave-boundary/src/untrusted.rs]: I think that's broken, you shouldn't be looking at `retval` unless the `result` is successful.
I think the right approach would be to go back to the original match statement and do something like this:
```
        match unsafe {
            ecall_fcn(
                eid,
                &mut retval,
                inbuf.as_ptr(),
                inbuf.len(),
                outbuf.as_mut_ptr(),
                outbuf.len(),
                &mut outbuf_used,
                &mut outbuf_retry_id,
            )
        } {
            sgx_status_t::SGX_SUCCESS => match retval {
                sgx_status_t::SGX_ERROR_OUT_OF_MEMORY => continue,
                sgx_status_t::SGX_SUCCESS => {
                    outbuf.truncate(outbuf_used);
                    break Ok(outbuf);
                }
                other_retval => {
                    if is_fatal_sgx_status(other_retval) {
                        panic!(...);
                    }
                    break (Err(other_retval)),
                }
            },
            status => {
                    if is_fatal_sgx_status(status) {
                        panic!(...);
                    }
                 break Err(status),
            }
 ```

File: enclave-boundary/src/untrusted.rs [hunk 3]
```diff
@@ -46,7 +46,20 @@ pub fn make_variable_length_ecall(
...
+        }
+        if is_fatal_sgx_status(result) {
+            panic!(
+                "Enclave reported fatal error: ecall returned {:?}. Panicking to restart.",
+                result
+            );
+        }
+        match result {
             sgx_status_t::SGX_SUCCESS => match retval {
                 sgx_status_t::SGX_ERROR_OUT_OF_MEMORY => continue,
```
Comment (nick-mobilecoin) [inline enclave-boundary/src/untrusted.rs:64]: If the "pre-check" is done, it needs to look at `result` first. If `result` fails we don't know what `retval` will have in it, it's up to the SDK if it overwrites the SGX_SUCCESS we put in, or if it's going to set it to some other value prior to making the actual enclave call. 

Github comment limitation :(
Suggest keeping the match statement and leveraging [Match Guards](https://doc.rust-lang.org/reference/expressions/match-expr.html#match-guards)
```rust
        match result {
            sgx_status_t::SGX_SUCCESS => match retval {
                sgx_status_t::SGX_ERROR_OUT_OF_MEMORY => continue,
                sgx_status_t::SGX_SUCCESS => {
                    outbuf.truncate(outbuf_used);
                    break Ok(outbuf);
                }
                other_retval if is_fatal_sgx_status(other_retval) => panic!("Enclave reported fatal error: ecall retval: {other_retval:?}. Panicking to restart."),
                other_retval => break (Err(other_retval)),
            },
            status if is_fatal_sgx_status(status) => panic!("Enclave reported fatal error: ecall returned {status:?}. Panicking to restart."),
            status => break Err(status),
        }
```

**Review by nick-mobilecoin — CHANGES_REQUESTED** [review]

File: enclave-boundary/src/untrusted.rs (hunk 3 above)
Comment (eranrund) [reply enclave-boundary/src/untrusted.rs:64]: Ooh yeah, match guards! That's more elegant than what I suggested :)

Comment (dolanbernard) [reply enclave-boundary/src/untrusted.rs:64]: Oh, neat, I didn't know about this feature

File: enclave-boundary/src/untrusted.rs [hunk 4]
```diff
@@ -46,16 +46,42 @@ pub fn make_variable_length_ecall(
...
                 sgx_status_t::SGX_SUCCESS => {
                     outbuf.truncate(outbuf_used);
                     break Ok(outbuf);
                 }
+                other_retval if is_fatal_sgx_status(other_retval) => {
+                    panic!(
+                        "Enclave reported fatal error: ecall retval: {:?}. Panicking to restart.",
```
Comment (eranrund) [inline enclave-boundary/src/untrusted.rs:59]: Nit but the restart behavior is a side-effect of how we run this in our k8 setup, it is not behavior that is guaranteed.

```suggestion
                        "Enclave reported fatal error: ecall retval: {:?}",
```

File: enclave-boundary/src/untrusted.rs [hunk 5]
```diff
@@ -46,16 +46,42 @@ pub fn make_variable_length_ecall(
...
+                        retval
+                    );
+                }
                 other_retval => break (Err(other_retval)),
             },
+            status if is_fatal_sgx_status(status) => {
+                panic!(
+                    "Enclave reported fatal error: ecall returned {:?}. Panicking to restart.",
```
Comment (eranrund) [inline enclave-boundary/src/untrusted.rs:67]: Nit but the restart behavior is a side-effect of how we run this in our k8 setup, it is not behavior that is guaranteed.

```suggestion
                    "Enclave reported fatal error: ecall returned {:?}",
```

**Review by eranrund — APPROVED** [review]

File: enclave-boundary/src/untrusted.rs [hunk 6]
```diff
@@ -46,16 +46,42 @@ pub fn make_variable_length_ecall(
...
             sgx_status_t::SGX_SUCCESS => match retval {
                 sgx_status_t::SGX_ERROR_OUT_OF_MEMORY => continue,
                 sgx_status_t::SGX_SUCCESS => {
                     outbuf.truncate(outbuf_used);
                     break Ok(outbuf);
                 }
+                other_retval if is_fatal_sgx_status(other_retval) => {
+                    panic!(
+                        "Enclave reported fatal error: ecall retval: {:?}",
+                        retval
```
Comment (nick-mobilecoin) [inline enclave-boundary/src/untrusted.rs:60]: Should probably use the local variable name
```suggestion
                        other_retval
```

File: enclave-boundary/src/untrusted.rs [hunk 7]
```diff
@@ -46,16 +46,42 @@ pub fn make_variable_length_ecall(
...
+                    );
+                }
                 other_retval => break (Err(other_retval)),
             },
+            status if is_fatal_sgx_status(status) => {
+                panic!(
+                    "Enclave reported fatal error: ecall returned {:?}",
+                    result
```
Comment (nick-mobilecoin) [inline enclave-boundary/src/untrusted.rs:68]: Similar to above consider local variable name
```suggestion
                    status
```

**Review by nick-mobilecoin — APPROVED** [review]

**Pull request merged**